### PR TITLE
Created option --watch for run plugin

### LINKED ISSUE #4064: Serverless Run: Add a file watching
# This is a Feature Proposal

## Description
Add a `--watch` flag for the new `serverless run` command. This flag will cause the source code and serverless.yml to be file watched. Any time a change is detected, the change should automatically be deployed to the local-emulator/event-gateway.

### PR DESCRIPTION
**Pull request by guidr**
## What did you implement:
I have implemented a `--watch` option to the `sls run` command. When changes are detected on the code, it will re-deploy the affected function(s) to the emulator.

Closes #4064

## How did you implement it:
I have created a new util function called `watchChanges` on the `run` plugin. It creates a watcher for each function defined on `serverless.yml`. This watcher will detect changes on the handler file or any recursive dependencies of it. I’ve used `node-precinct` to find the dependencies of the handler file and `node-watch` to watch the files.

## How can we verify it:
The new option is available on `sls run` command:

`sls run --watch`
or the shortcut: `sls run -w`

When changes are detected, the console will produce the following messages:

Handler file changed:
```
Serverless     Function 'users-ListUsers' update triggered by change on file functions/listUsers.js
Serverless     Function 'users-ListUsers' loaded
```

A dependency of the handler file:
```
Serverless     Function 'users-ListUsers' update triggered by change on file src/utils.js
Serverless     Function 'users-ListUsers' loaded
```

A dependency of multiple functions:
```
Serverless     Function 'users-ListUsers' update triggered by change on file src/models/User.js
Serverless     Function 'users-CreateUser' update triggered by change on file src/models/User.js
Serverless     Function 'users-ListUsers' loaded
Serverless     Function 'users-CreateUser' loaded
```

## Todos:

- [x] Write tests
- [ ] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by RafalWilinski** [comment]
Thanks @guidr for this Pull Request! 🙌 

It would be awesome if you could also write some tests for this functionality.

File: package.json [hunk 1]
```diff
@@ -114,7 +114,9 @@
     "moment": "^2.13.0",
     "node-fetch": "^1.6.0",
     "node-forge": "^0.7.1",
+    "node-watch": "^0.5.5",
     "opn": "^5.0.0",
+    "precinct": "^3.8.0",
```
Comment (RafalWilinski) [inline package.json]: Could you also please add `package-lock.json` file?

File: lib/plugins/run/utils/watchChanges.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      deps,
+      {
+        recursive: false,
+        filter: file => !/node_modules/.test(file),
+      },
+      (event, filename) => {
+        const fileLocation = filename.replace(`${servicePath}/`, '');
+
+        // eslint-disable-next-line max-len
+        logServerless(`Function '${service.service}-${functionName}' update triggered by change on file ${fileLocation}`);
```
Comment (RafalWilinski) [inline lib/plugins/run/utils/watchChanges.js]: I suggest something like:
```
logServerless([
  `Function '${service.service}-${functionName}' update triggered`,
  `by change on file ${fileLocation}`
].join(''));
```

so we could get rid of `eslint-disable-next-line max-len`

File: lib/plugins/run/utils/watchChanges.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    const handlerFile = path.basename(functionConfig.handler, handlerName);
+    const handlerPath = path.dirname(functionConfig.handler);
+
+    const deps = getDependencies(path.join(servicePath, `${handlerPath}.js`, handlerFile));
+
+    watch(
+      deps,
+      {
+        recursive: false,
+        filter: file => !/node_modules/.test(file),
```
Comment (RafalWilinski) [inline lib/plugins/run/utils/watchChanges.js]: I think you can also exclude directories like `.serverless`.

Furthermore, I'm wondering whether this should be configurable or taken straight from `package` configuration section from `serverless.yml`.

Comment (guidr) [reply lib/plugins/run/utils/watchChanges.js]: To be honest, this filter is not required in here as `precinct` will ignore any core dependencies. The same applies for other directories like `.serverless`. Only dependencies included on the handler files will be watched (those included via `require()` or `import`)

**Comment by guidr** [comment]
Thank you for the suggestions @RafalWilinski. I'll add a few commits to this PR to address your comments and also to add the unit tests.

File: package.json (hunk 1 above)
Comment (guidr) [reply package.json]: @RafalWilinski this has now been added

**Comment by pmuens** [comment]
Thank you so much for your work on this @guidr 👍 💯 🥇

Overall the functionality and the value you've added is really great! Unfortunately `serverless run` is not actively maintained at the moment. I'll close this PR for now, but we'll definitely re-visit it once we plan to implement functionality like that for other commands.

Thanks again for working on this!

**Pull request closed without merging**